### PR TITLE
[AMQP] Ensure ReceiverEndpointState completes in case no link remote-close ack

### PR DESCRIPTION
**Pull request by anuchandy**
Refer the git_ticket https://github.com/Azure/azure-sdk-for-java/issues/29212 for details.

### CODE REVIEW

**Comment by anuchandy** [comment]
This is a draft, doing some EtoE testing, may update pr based on EtoE test observation

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 1]
```diff
@@ -184,7 +188,9 @@ protected ReactorReceiver(AmqpConnection amqpConnection, String entityPath, Rece
 
     @Override
     public Flux<AmqpEndpointState> getEndpointStates() {
-        return endpointStates.distinct();
+        return endpointStates
+            .distinct()
```
Comment (conniey) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:192]: I believe it should be `distinctUntilChanged()`. `distinct` keeps a set, so if you see INACTIVE -> ACTIVE -> INACTIVE states, it'll only output INACTIVE and ACTIVE but not the last state.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java [hunk 2]
```diff
@@ -282,6 +288,15 @@ protected Message decodeDelivery(Delivery delivery) {
...
      * </p>
      *
+     * <p>
+     * Closing ReactorReceiver involves 3 stages, running in following order  -
+     * <ul>
+     *      <li>local-close (client to broker) via beginClose() </li>
+     *      <li>remote-close ack (broker to client)</li>
+     *      <li>disposal of ReactorReceiver resources via completeClose()</li>
+     * </ul>
+     * Ref:https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-amqp/docs/reactor-receiver-closeflow.png
```
Comment (conniey) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:298]: nit: Replace with `@link <a href="">Reactor receiver close flow</a>`

**Review by conniey — APPROVED** [review]
Thanks for this! I was also wondering if we'd have to check for the same pattern in ReactorSender since they are similar.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java (hunk 1 above)
Comment (anuchandy) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java:192]: ah, I see, good point! will switch to `distinctUntilChanged` .

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java [hunk 3]
```diff
@@ -82,14 +82,17 @@ public class AmqpReceiveLinkProcessor extends FluxProcessor<AmqpReceiveLink, Mes
...
-     * @param parentConnection Represents the parent connection.
+     * @param connectionProcessor A {@link Disposable} reference to the connection-processor which produces
+     *                           the connection on which receive-links are hosted. It is used to eagerly check
+     *                           if the connection-processor is disposed so that this link-processor can self dispose.
      *
      * @throws NullPointerException if {@code retryPolicy} is null.
      * @throws IllegalArgumentException if {@code prefetch} is less than 0.
      */
-    public AmqpReceiveLinkProcessor(String entityPath, int prefetch, Disposable parentConnection) {
+    public AmqpReceiveLinkProcessor(String entityPath, int prefetch, Disposable connectionProcessor) {
```
Comment (anuchandy) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java:92]: The name `parentConnection` could be misleading, as it's not the disposal of amqp-connection being observed but the connection-processor.

**Comment by anuchandy** [comment]
> Thanks for this! I was also wondering if we'd have to check for the same pattern in ReactorSender since they are similar.

Yes we need similar pattern for sender as well, will work on `ReactorSender` EndpointState termination in a separate pr;

**Pull request merged**